gio/src/subclass/async_initable.rs

```diff
@@ -0,0 +1,390 @@
+                user_data: glib::ffi::gpointer,
+            ) where
+                T: AsyncInitableImpl,
+            {
+                let type_data = T::type_data();
+                let parent_iface = type_data.as_ref().parent_interface::<AsyncInitable>()
+                    as *const ffi::GAsyncInitableIface;
+                let init_finish = (*parent_iface)
+                    .init_finish
+                    .expect("no parent \"init_finish\" implementation");
+
+                let r: Box<ThreadGuard<GioFutureResult<(), Error>>> =
+                    Box::from_raw(user_data as *mut _);
+                let r = r.into_inner();
+
+                let mut error = ptr::null_mut();
+                init_finish(source_object as *mut _, res, &mut error);
+                let result = if error.is_null() {
+                    Ok(())
+                } else {
+                    Err(from_glib_full(error))
+                };
+                r.resolve(result);
+            }
+
+            Box::pin(crate::GioFuture::new(
+                initable,
+                move |obj, cancellable, res| {
+                    let user_data: Box<ThreadGuard<GioFutureResult<(), Error>>> =
+                        Box::new(ThreadGuard::new(res));
+                    let user_data = Box::into_raw(user_data);
+                    init_async(
+                        obj.unsafe_cast_ref::<AsyncInitable>().to_glib_none().0,
+                        io_priority.into_glib(),
+                        cancellable.to_glib_none().0,
+                        Some(parent_init_future_callback::<T>),
+                        user_data as *mut _,
+                    );
+                },
+            ))
+        }
+    }
+}
+
+unsafe impl<T: AsyncInitableImpl> IsImplementable<T> for AsyncInitable {
+    fn interface_init(iface: &mut glib::Interface<Self>) {
+        let iface = iface.as_mut();
+        iface.init_async = Some(async_initable_init_async::<T>);
+        iface.init_finish = Some(async_initable_init_finish::<T>);
+    }
+}
+
+unsafe extern "C" fn async_initable_init_async<T: AsyncInitableImpl>(
+    initable: *mut ffi::GAsyncInitable,
+    io_priority: std::os::raw::c_int,
+    cancellable: *mut ffi::GCancellable,
+    callback: ffi::GAsyncReadyCallback,
+    user_data: glib::ffi::gpointer,
+) {
+    let instance = &*(initable as *mut T::Instance);
+    let imp = instance.imp();
+    let initable = from_glib_borrow::<_, AsyncInitable>(initable);
+    let cancellable = from_glib_borrow::<_, Option<Cancellable>>(cancellable);
+
+    let task = callback.map(|callback| {
+        let task = LocalTask::new(
+            Some(initable.upcast_ref::<glib::Object>()),
+            cancellable.as_ref().as_ref(),
+            move |task, obj| {
+                let result: *mut crate::ffi::GAsyncResult =
+                    task.upcast_ref::<AsyncResult>().to_glib_none().0;
+                let obj: *mut glib::object::GObject = obj.to_glib_none().0;
```

Comment:
The cast and the type annotation shouldn't be needed here. How does it fail without?

Reply:
Feel free to keep the type annotation for documentation purposes, I'm just wondering if there's something wrong here

Reply:
I don't know why this is here, I just copied it from somewhere else. The type ascription isn't necessary and it works without it